### PR TITLE
fix mbedtls_x509_dn_gets to show non-ASCII string properly

### LINKED ISSUE #3303: mbedtls_x509_crt_info cannot show non-ASCII string properly
### Description
- Type: Bug 
- Priority: Minor

---------------------------------------------------------------
## Bug

**OS**  
macOS Majove

**mbed TLS build:**  
Version: git-d9f694960
OS version: 10.14.6 
Compiler and options (if you used a pre-built binary, please indicate how you obtained it): 
mbedtls is installed as follows:

```
$ git clone https://github.com/ARMmbed/mbedtls.git
$ cd mbedtls
$ git co d9f694960
$ mkdir build && cd build
$ cmake ../
$ cmake --build .
$ sudo make install
```

Then, compile the source codes contained in the attached zip file ([sample.zip](https://github.com/ARMmbed/mbedtls/files/4578796/sample.zip)).

```
$ unzip sample.zip
$ cd sample && make
g++ -std=c++14 main.cpp -lmbedx509 -lmbedtls -lmbedcrypto -o a.out
$ ./a.out
```

This program generates `output.txt`. Open `output.txt` and see the "subject name" row.

```
subject name      : C=PL, O=Tomasz MoÅ?, CN=Tomasz MoÅ?, emailAddress=desowin@gmail.com
                                     ^^              ^^
```

`O=Tomasz Moń, CN=Tomasz Moń` is correct, but `mbedtls_x509_crt_info` fails to encode non-ASCII character `ń` and "subject name" is not properly shown.

I investigated this problem and I found the following lines are the root of the cause.

https://github.com/ARMmbed/mbedtls/blob/d9f694960f10c23cc5d7cb6de440fe3584ff7979/library/x509.c#L783-L793

`ń` is encoded as `0xc5 0x84` in UTF-8. The above lines convert `0xc5 0x84` into `0xc5 0x3f`, but this byte sequence can't be interpreted as UTF-8 encoded string. Therefore, the "subject name" row is not shown properly.

**Expected behavior**   
If `mbedtls_x509_crt_info` should escape non-ASCII string, the output should be

```
subject name      : C=PL, O=Tomasz Mo??, CN=Tomasz Mo??, emailAddress=desowin@gmail.com
```

### PR DESCRIPTION
**Pull request by kohnakagawa**
## Description
This pull-request fixes #3303.

## Status
**READY**

### CODE REVIEW

File: library/x509.c [hunk 1]
```diff
@@ -787,7 +787,7 @@ int mbedtls_x509_dn_gets( char *buf, size_t size, const mbedtls_x509_name *dn )
                 break;
 
             c = name->val.p[i];
-            if( c < 32 || c == 127 || ( c > 128 && c < 160 ) )
+            if( c < 32 || c == 127 || ( c > 128) )
```
Comment (danh-arm) [inline library/x509.c:790]: Could be simplified to `if( c < 32 || c >=127)

Comment (yanesca) [reply library/x509.c:790]: Also, please note that the [Mbed TLS Coding Standard](https://tls.mbed.org/kb/development/mbedtls-coding-standards) requires a space character before closing brackets.

**Comment by kohnakagawa** [comment]
Thank you for your reply. I added ChangeLog.d and fixed to follow Mbed TLS Coding Standard.

**Comment by kohnakagawa** [comment]
@danh-arm
Sorry, I squashed these commits together and clean up the commit log.

**Comment by kohnakagawa** [comment]
Some checks were not successful, but I cannot figure out why these checks failed because I cannot see the details of these failures.

In my local environment (macOS Majove, clang version 11.0.0), mbedTLS was successfully built, and all tests were passed.

Could you please tell me why these checks were failed?

**Comment by gilles-peskine-arm** [comment]
The test failure looks like some glitch on the CI. One test case in `ssl-opt.sh` failed, only in one configuration (clang on Debian 9 on x86_64). The same test case passed in other configurations and in the merge. The failing test case is `Session resume using tickets, DTLS: openssl client`. Logs: 
[std-make-full-config-clang-debian-9-x64-o-cli-163.log](https://github.com/ARMmbed/mbedtls/files/4649179/std-make-full-config-clang-debian-9-x64-o-cli-163.log)
[std-make-full-config-clang-debian-9-x64-o-srv-163.log](https://github.com/ARMmbed/mbedtls/files/4649182/std-make-full-config-clang-debian-9-x64-o-srv-163.log)

This isn't an intermittent failure and I don't see an obvious cause. I've restarted the pr-head job to see if it's reproducible.

File: ChangeLog.d/fix-print-non-ascii-string-in-mbedtls_x509_crt_info.txt [hunk 2]
```diff
@@ -0,0 +1,3 @@
+Changes
+   * Fix mbedtls_x509_crt_infoto show non-ASCII string properly.
```
Comment (gilles-peskine-arm) [inline ChangeLog.d/fix-print-non-ascii-string-in-mbedtls_x509_crt_info.txt:2]: Non-ASCII characters are replaced by `?`, so I don't think this description is correct.

Comment (kohnakagawa) [reply ChangeLog.d/fix-print-non-ascii-string-in-mbedtls_x509_crt_info.txt]: You're right. I fixed this description.

**Comment by kohnakagawa** [comment]
@gilles-peskine-arm
> Please add a non-regression test.
OK. I will create a non-regression test and again push it.

**Comment by kohnakagawa** [comment]
I added a non-regression test and changed ChangeLog.
Could you please review again?

**Comment by gilles-peskine-arm** [comment]
Given that this is a minor bug in a function that is for debugging and logging purposes, I don't think the fix needs to be backported.

**Comment by gstrauss** [comment]
@gilles-peskine-arm I think that the fix here breaks the display of valid UTF-8 outside of ASCII chars.

UTF-8 is valid in distinguished names.  I think that the original code `if( c < 32 || c == 127 || ( c > 128 && c < 160 ) )` may have **incorrectly** been attempting to exclude overlong UTF-8 sequences by checking `( c > 128 && c < 160 )`, but that check is insufficient to detect a multi-byte overlong UTF-8 encoding, generating a false positive for the 0xc5 0x84 example in #3303.  (A better check would reject 0xc0 and 0xc1)

In any case, the change in this PR replaces all non-ASCII with '?'.  While that makes for a consistent display, it also leads to valid non-ASCII UTF-8 replaced with all '?'.  Maybe instead of `if( c < 32 || c >= 127 )`, only CTLs should be rejected:  i.e. `if( c < 32 || c == 127 )`.  `mbedtls_x509_dn_gets()` is a function used to generate strings to display and so CTLs should continue to be replaced with '?' because they were historically replaced by '?' in `mbedtls_x509_dn_gets()` and some existing usage might depend on this behavior.

(The reason for revisiting this is that I am looking at performance and correctness of `mbedtls_x509_dn_gets()` related to #5431)

**Comment by gilles-peskine-arm** [comment]
@gstrauss This patch changed `dn_gets` from passing most non-ASCII bytes unchanged to only passing printable ASCII through. The intent was to stop emitting invalid UTF-8 no matter what was in the certificate.

The previous code with `c > 128 && c < 160` was already present in XySSL 0.1 in 2009.  My guess is that it was intended for latin1, not UTF-8. (And it should have been `c >= 128 && c < 160`.)

Since Mbed TLS is intended for environments with limited resources, it can't realistically be fully correct with respect to text encoding. (We aren't going to do Unicode normalization!) 

I'm not sure what the ideal behavior is for `mbedtls_x509_dn_gets`. In this PR we chose to make the output non-disruptive (ASCII only), but it's also non-informative. I think that's fine for `mbedtls_x509_crt_info`, which is mainly for debugging purposes, but not for `mbedtls_x509_dn_gets`, which applications can use for other purposes.

If lighttpd uses the output of `mbedtls_x509_dn_gets` for other purposes, and you want to preserve non-ASCII characters, can you please open a separate feature request for this, and explain your constraints (with respect to diversity of certificates to support, acceptability of invalid output, etc.)?

**Comment by gstrauss** [comment]
> If lighttpd uses the output of `mbedtls_x509_dn_gets` for other purposes, and you want to preserve non-ASCII characters, can you please open a separate feature request for this, and explain your constraints (with respect to diversity of certificates to support, acceptability of invalid output, etc.)?

Follow-up in #5431

**Pull request merged**